### PR TITLE
PBKDF2 implementation

### PR DESCRIPTION
**Pull request by lsolodkova**
PBKDF2 is implemented as external function together with one auxiliary function for XOR'ing byte arrays. Should I make'em class methods?

### CODE REVIEW

**Comment by mvondracek** [comment]
What's the status?
// @lsolodkova, @sobuch

**Comment by lsolodkova** [comment]
Well, what status? I've replaced library function with manual implementation and added password length check.

**Comment by mvondracek** [comment]
Great! Is it tested? Will it be ready for merge to dev once it has tests?


26. 10. 2019 v 9:56, lsolodkova <notifications@github.com>:

File: PA193_mnemonic_Slytherin/mnemonic.py [hunk 1]
```diff
@@ -25,6 +26,33 @@
...
+    Uses HMAC-SHA512, derived key length is 512 bit
+    :rtype: bytes
+    :return: key derived by PBKDF2 algorithm
+    """
+    # The first iteration of PRF uses Password as the PRF key
+    # and Salt concatenated with i encoded as a big-endian
+    # 32-bit integer as the input.
+    U = hmac.new(passw, salt + (1).to_bytes(4, byteorder='big'), digestmod=sha512).digest()
+    F = U  # F is the xor (^) of c iterations of chained PRFs
+    for i in range(1, rounds):
```
Comment (mvondracek) [inline PA193_mnemonic_Slytherin/mnemonic.py]: > he iteration count is set to 2048
> https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#from-mnemonic-to-seed

`range(1, 2048)` is 2047 iterations. `len(range(1, 2048)) == 2047`

**Review by mvondracek — CHANGES_REQUESTED** [review]

File: PA193_mnemonic_Slytherin/mnemonic.py (hunk 1 above)
Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: First iteration is  U = hmac.new(passw, salt + (1).to_bytes(4, byteorder='big'), digestmod=sha512).digest(), it is computed manually, [U1 = PRF(Password, Salt + INT_32_BE(i))](https://en.wikipedia.org/wiki/PBKDF2#Key_derivation_process)

Comment (lsolodkova) [reply PA193_mnemonic_Slytherin/mnemonic.py]: I don't know if we have to generalize the implementation to the arbitrary key length, I've just hardcoded that.

Comment (mvondracek) [reply PA193_mnemonic_Slytherin/mnemonic.py]: Didn't recognize that, thanks for explaining! :)

I'm fine with `pbkdf2_sha512` with fixed key length. Outside of this project we would use functions for this from standard library, no need to implement general solution for that. :)

**Review by mvondracek — APPROVED** [review]

**Comment by mvondracek** [comment]
I'm happy to merge this. @sobuch, please have a look at it.

**Review by sobuch — APPROVED** [review]

**Pull request merged**